### ISSUE TITLE
Configuration cache problems

### ISSUE THREAD

**Issue by eygraber** [vanniktech/gradle-maven-publish-plugin#259]
Looks like there are some issues with using this plugin when the configuration cache is enabled. Not sure if any of this is internal to Gradle or part of the plugin, but figured I'd record it here.

[Here's a build of mine that failed because of it.](https://github.com/eygraber/kodein-di-android-components/runs/2635880947)

**Comment by gabrielittner** [comment]
I'll add support for it once the underlying Maven Publish and Signing plugins support configuration caching on the Gradle side https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:plugins

**Comment by gabrielittner** [comment]
`maven-publish` itself supports cc since 7.6 and the `signing` plugin will in Gradle 8.1. However there is one big open blocker when publishing to maven central https://github.com/gradle/gradle/issues/22779. When configuration cache is enabled the repository url will be evaluated eagerly and not when the publishing tasks are executed. Since we are computing that url based on the output of the task that creates the staging repo the build will fail with this:

```
Configuration cache state could not be cached: field `repository` of `org.gradle.api.publish.maven.tasks.PublishToMavenRepository$PublishSpec` bean found in field `value` of `org.gradle.internal.Try$Success` bean found in field `result` of `org.gradle.internal.serialization.Cached$Fixed` bean found in field `spec` of task `:nexus:publishMavenPublicationToMavenCentralRepository` of type `org.gradle.api.publish.maven.tasks.PublishToMavenRepository`: error writing value of type 'org.gradle.api.publish.maven.tasks.PublishToMavenRepository$RepositorySpec$Configured'
> Cannot query the value of this provider because it has no value available.
  The value of this provider is derived from:
    - task ':nexus:createStagingRepository' property 'stagingRepositoryId'
```

So current status: 
- Gradle 7.6: publishing to local or non maven central remote repositories works with cc when signing is disabled
- Gradle 8.1: publishing to local or non maven central remote repositories works with cc with signing enabled
- TBD: publishing to maven central works with cc

**Comment by gabrielittner** [comment]
With Gradle 8.1 being out the updated status is
- Publishing releases to Maven Central (snapshots are fine), blocked by [Gradle issue #22779](https://github.com/gradle/gradle/issues/22779).
- Kotlin Multiplatform projects, blocked by [KT-49933](https://youtrack.jetbrains.com/issue/KT-49933).

**Comment by sureshg** [comment]
Had to diable the configuration cache (`gradle-8.12`) as build was failing with the following error.  Is this issue still blocked by https://github.com/gradle/gradle/issues/22779 ?

```
Configuration cache state could not be cached: field `repository` of `org.gradle.api.publish.maven.tasks.PublishToMavenRepository$PublishSpec` bean found in field `value` of `org.gradle.internal.Try$Success` bean found in field `result` of `org.gradle.internal.serialization.Cached$Fixed` bean found in field `spec` of task `:catalog:publishMavenPublicationToMavenCentralRepository` of type `org.gradle.api.publish.maven.tasks.PublishToMavenRepository`: error writing value of type 'org.gradle.api.publish.maven.tasks.PublishToMavenRepository$RepositorySpec$Configured'
> Services of type BuildFeatures are not available for injection into instances of type BuildService.`
```

**Comment by eygraber** [comment]
https://github.com/gradle/gradle/issues/22779#issuecomment-2992618811

> The recommended approach for 9.0 is to mark a task that requires a repository URL to be defined during execution as notCompatibleWithConfigurationCache().

Sounds like it might be some time until there's actual support. Is it worth it to mark the task as `notCompatibleWithConfigurationCache` so we don't have to disable it on our projects manually?

**Comment by eygraber** [comment]
Actually, just saw https://github.com/vanniktech/gradle-maven-publish-plugin/pull/1006 which seems like it might remove the issue altogether?

**Comment by gabrielittner** [comment]
Yes with that PR we are compatible without any changes on Gradle's side needed. The ETA for that is early July (mostly waiting for the OSSRH shutdown at the end of this month)
